### PR TITLE
Implementation of `Identity` serivce

### PR DESCRIPTION
**Pull request by uselessgoddess**

### CODE REVIEW

**Comment by uselessgoddess** [comment]
Add the following later: 
- email confirmation 
- password reset
- 2fa auth
- access policies 
- logging and caching

File: Identity/Core/Models/RefreshToken.cs
Comment (MefistoSamael) [inline Identity/Core/Models/RefreshToken.cs:1]: move to domain layer

Comment (MefistoSamael) [reply Identity/Core/Models/RefreshToken.cs:1]: place empty lines between methods

File: Identity/Core/Models/AppUser.cs
Comment (MefistoSamael) [inline Identity/Core/Models/AppUser.cs:1]: move to domain

File: Identity/Core/Models/Role.cs
Comment (MefistoSamael) [inline Identity/Core/Models/Role.cs:1]: move to domain

File: Identity/Core/Commands/Login.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+public class LoginValidation : AbstractValidator<Login>
+{
+    public LoginValidation()
+    {
+        RuleFor(x => x.Username)
+            .NotEmpty();
+
+        RuleFor(x => x.Password)
+            .NotEmpty();
+    }
```
Comment (MefistoSamael) [inline Identity/Core/Commands/Login.cs]: separate all validators and DTO's

File: Identity/Core/Commands/Validation.cs
Comment (MefistoSamael) [inline Identity/Core/Commands/Validation.cs:1]: rename class. Move in root of core

File: Identity/Core/Handlers/DemoteHandler.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using MediatR;
+using Microsoft.AspNetCore.Identity;
+
+public class DemoteHandler(
+    UserManager<AppUser> users) : IRequestHandler<Demote>
+{
+    public async Task Handle(Demote req, CancellationToken cancellationToken)
+    {
+        var user = await users.FindByIdAsync(req.Id) ?? throw new NotFoundException();
+        await users.RemoveFromRolesAsync(user, req.Roles.Select(r => r.ToString()));
```
Comment (MefistoSamael) [inline Identity/Core/Handlers/DemoteHandler.cs]: await must be preceeded and followed by blank lines

File: Identity/Core/Handlers/LoginHandler.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    DatabaseCx cx) : IRequestHandler<Login, TokensPair?>
+{
+    public async Task<TokensPair?> Handle(Login req, CancellationToken cancellationToken)
+    {
+        var result = await sign.PasswordSignInAsync(req.Username, req.Password, isPersistent: false,
+            lockoutOnFailure: false);
+
+        if (!result.Succeeded)
+        {
+            throw new UnauthorizedException("invalid username or password");
```
Comment (MefistoSamael) [inline Identity/Core/Handlers/LoginHandler.cs]: it is badrequest

File: Identity/Core/Handlers/RefreshHandler.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public async Task<TokensPair?> Handle(Refresh req, CancellationToken cancellationToken)
+    {
+        var old =
+            await cx.RefreshTokens.FirstOrDefaultAsync(rt => rt.Token == req.Token,
+                cancellationToken: cancellationToken);
+
+        if (old == null || old.Expire < DateTime.Now || old.IsRevoked)
+        {
+            throw new UnauthorizedException();
```
Comment (MefistoSamael) [inline Identity/Core/Handlers/RefreshHandler.cs]: use badrequest exception

File: Identity/Core/Handlers/RegisterHandler.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+            throw new ValidationException(new[]
+                { new ValidationFailure("Email", "user with the same email already exists") });
+        }
+
+        var user = new AppUser
+            { UserName = req.Username, Email = req.Email, CreatedDate = DateTime.UtcNow };
+        if (!(await users.CreateAsync(user, req.Password)).Succeeded)
+        {
+            throw new BadRequestException("user already exists");
+        }
```
Comment (MefistoSamael) [inline Identity/Core/Handlers/RegisterHandler.cs]: 1. conditional statements must be followed and preceeded by blank lines

2. Use alreadyExists exception

File: Identity/Core/Handlers/RegisterHandler.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.AspNetCore.Identity;
+
+public class RegisterHandler(UserManager<AppUser> users) : IRequestHandler<Register>
+{
+    public async Task Handle(Register req, CancellationToken cancellationToken)
+    {
+        if (await users.FindByEmailAsync(req.Email) == null)
+        {
+            throw new ValidationException(new[]
+                { new ValidationFailure("Email", "user with the same email already exists") });
```
Comment (MefistoSamael) [inline Identity/Core/Handlers/RegisterHandler.cs]: use badRequest exception

File: Identity/Core/QueryHandlers/UsersHandlers.cs
Comment (MefistoSamael) [inline Identity/Core/QueryHandlers/UsersHandlers.cs:1]: 1. separate handlers into different files
2. use more descriptive names for handlers

File: Identity/Infrastructure/Data/DatabaseCx.cs
Comment (MefistoSamael) [inline Identity/Infrastructure/Data/DatabaseCx.cs:1]: use more descriptive name

File: Identity/Infrastructure/Utils.cs
Comment (MefistoSamael) [inline Identity/Infrastructure/Utils.cs:1]: seed db with HasData method

File: Identity/Program.cs
Comment (MefistoSamael) [inline Identity/Program.cs:1]: separate services configuration in different files by layers

**Review by MefistoSamael — CHANGES_REQUESTED** [review]
Each layer of application has to be created as class library(except presentation)

File: Identity/Core/Commands/Login.cs (hunk 1 above)
Comment (uselessgoddess) [reply Identity/Core/Commands/Login.cs]: but this isn't DTO))

File: Identity/Core/Handlers/DemoteHandler.cs (hunk 2 above)
Comment (uselessgoddess) [inline Identity/Core/Handlers/DemoteHandler.cs]:
```suggestion
        var user = await users.FindByIdAsync(req.Id) ?? throw new NotFoundException();

        await users.RemoveFromRolesAsync(user, req.Roles.Select(r => r.ToString()));
```

File: Identity/Core/Handlers/LoginHandler.cs (hunk 3 above)
Comment (uselessgoddess) [inline Identity/Core/Handlers/LoginHandler.cs]:
```suggestion
            throw new BadRequestException("invalid username or password");
```

File: Identity/Core/Handlers/RefreshHandler.cs (hunk 4 above)
Comment (uselessgoddess) [inline Identity/Core/Handlers/RefreshHandler.cs]:
```suggestion
            throw new BadRequestException();
```

File: Identity/Core/Handlers/RegisterHandler.cs (hunk 6 above)
Comment (uselessgoddess) [inline Identity/Core/Handlers/RegisterHandler.cs]:
```suggestion
            throw new BadRequestException("user with the same email already exists");
```

File: Identity/Core/QueryHandlers/UsersHandlers.cs
Comment (uselessgoddess) [reply Identity/Core/QueryHandlers/UsersHandlers.cs:1]: 1. ok
2. names are subjective, but their len saves the memory

File: Identity/Infrastructure/Data/DatabaseCx.cs
Comment (uselessgoddess) [reply Identity/Infrastructure/Data/DatabaseCx.cs:1]: how can I improve descriptive of `Database Context`

File: Identity/Infrastructure/Utils.cs
Comment (uselessgoddess) [reply Identity/Infrastructure/Utils.cs:1]: but why? my way is easier and hides low-level representation

File: Identity/Core/QueryHandlers/UsersHandlers.cs
Comment (MefistoSamael) [reply Identity/Core/QueryHandlers/UsersHandlers.cs:1]: you shoudn't care so much about memory in such types of application

Comment (MefistoSamael) [reply Identity/Core/QueryHandlers/UsersHandlers.cs:1]: at first you should care about maintainability and extendibility

File: Identity/Infrastructure/Utils.cs
Comment (MefistoSamael) [reply Identity/Infrastructure/Utils.cs:1]: naming conventions) ¯\_(ツ)_/¯

File: Identity/Api/Program.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                Reference = new OpenApiReference
+                {
+                    Type = ReferenceType.SecurityScheme,
+                    Id = "Bearer",
+                },
+            },
+            []
+        },
+    });
+});
```
Comment (MefistoSamael) [inline Identity/Api/Program.cs]: take it away from program

File: Identity/Api/App/Controllers/AccountController.cs
Comment (MefistoSamael) [inline Identity/Api/App/Controllers/AccountController.cs:1]: separate business logic from presentation logic

**Review by MefistoSamael — CHANGES_REQUESTED** [review]

File: Identity/Api/Program.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+var config = builder.Configuration;
+
+builder.Services.AddInfrastructure(config);
+builder.Services.AddUseCases();
+builder.AddApi();
+
+builder.Services.AddControllers();
+builder.Services.AddDataValidation();
+
+builder.Services.AddSwaggerGen(c => { c.AddSecurityDefinition("Bearer", new OpenApiSecurityScheme()); });
```
Comment (MefistoSamael) [inline Identity/Api/Program.cs]: move to api configuration

File: Identity/Api/DependencyInjection.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                    ValidIssuer = jwt.GetValue<string>("Issuer"),
+                    ValidAudience = jwt.GetValue<string>("Audience"),
+                    IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(jwt.GetValue<string>("Key")!)),
+                };
+            });
+
+        builder.Services.AddAuthorizationBuilder()
+            .AddPolicy(Policy.Admin, policy => policy.RequireRole("Admin"));
+
+        return builder;
```
Comment (MefistoSamael) [inline Identity/Api/DependencyInjection.cs]: move to infrastructure configuration

File: Identity/UseCases/Services/TokenService.cs
Comment (MefistoSamael) [inline Identity/UseCases/Services/TokenService.cs:1]: move to infrastructure

**Review by MefistoSamael — CHANGES_REQUESTED** [review]

**Review by MefistoSamael — APPROVED** [review]

**Pull request merged**